File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.head = null;
    this.rear = null;
    this.length = 0;
  }

  enqueue(data) {
    const node = new Node(data);
    if (!this.head) {
      this.head = node;
    } else {
      this.rear.next = node;
    }
    this.rear = node;
    this.length++;
  }

  dequeue() {
    if (!this.head) {
      return false;
    }
    const data = this.head.data;
    this.head = this.head.next;
    this.length--;

    return data;
  }
}

const solution = (input) => {
  const [rowN, colN] = input[0].split(" ").map((v) => Number(v));
  const grid = [];

  for (let i = 1; i < input.length - 1; i++) {
    grid.push(input[i].split(" ").map((v) => Number(v)));
  }
  const [height, width, Sr, Sc, Fr, Fc] = input[input.length - 1]
    .split(" ")
    .map((v) => Number(v));

  const [startX, startY] = [Sr - 1, Sc - 1];
  const [targetX, targetY] = [Fr - 1, Fc - 1];

  const dirs = [
    [-1, 0],
    [1, 0],
    [0, 1],
    [0, -1],
  ];
  const visited = Array(rowN)
    .fill(0)
    .map((v) => Array(colN).fill(false));

  // const queue = [[startX, startY, 0]];
  const queue = new Queue();
  queue.enqueue([startX, startY, 0]);
  visited[startX][startY] = true;

  while (queue.length) {
    const [x, y, temp] = queue.dequeue();

    if (x === targetX && y === targetY) {
      console.log(temp);
      return;
    }

    for (let [dx, dy] of dirs) {
      const nextX = x + dx;
      const nextY = y + dy;

      const isValid =
        nextX >= 0 &&
        nextX < rowN &&
        nextX + height - 1 < rowN &&
        nextY >= 0 &&
        nextY < colN &&
        nextY + width - 1 < colN &&
        !visited[nextX][nextY];

      if (isValid) {
        visited[nextX][nextY] = true;

        let isNoWall = true;
        for (let i = nextX; i < nextX + height; i++) {
          for (let j = nextY; j < nextY + width; j++) {
            if (grid[i][j] === 1) {
              isNoWall = false;
            }
          }
        }

        if (isNoWall) {
          queue.enqueue([nextX, nextY, temp + 1]);
        }
      }
    }
  }

  console.log(-1);
};

solution(input);

File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.py
from collections import deque

rowN, colN = list(map(int, input().split()))
grid = []

for i in range(rowN):
  grid.append(list(map(int, input().split())))

sizeH, sizeW, startX, startY, finishX, finishY = list(map(int, input().split()))
visited = [[False]*colN for _ in range(rowN)]
visited[startX-1][startY-1] = True

queue = deque()
queue.append([startX-1, startY-1, 0])
answer = -1

def checkWall(x, y):
  if x + sizeH-1 >= rowN or y + sizeW-1 >= colN:
    return False

  for i in range(x, x+sizeH):
    if grid[i][y] == 1 or grid[i][y+sizeW-1]:
      return False

  for j in range(y, y+sizeW):
    if grid[x][j] == 1 or grid[x+sizeH-1][j]:
      return False

  return True

dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

while len(queue):
  x, y, cnt = queue.popleft()

  if x == finishX-1 and y == finishY-1:
    answer = cnt
    break

  for dx, dy in dirs:
    nextX = x + dx
    nextY = y + dy
    isValid = 0<=nextX<rowN and 0<=nextY<colN and not visited[nextX][nextY] and grid[nextX][nextY] == 0

    if isValid:
      visited[nextX][nextY] = True
      if checkWall(nextX, nextY):
        queue.append([nextX, nextY, cnt + 1])

print(answer)